File: packages/sdk-js/src/normalize.ts
import type { CommonFields, Json, JsonObject } from './types.js'

const fieldMap: Record<string, string> = {
  tenantId: 'tenant_id',
  serviceNamespace: 'service.namespace',
  serviceInstanceId: 'service.instance.id',
  serviceVersion: 'service_version',
  hostName: 'host.name',
  hostId: 'host.id',
  traceId: 'trace_id',
  spanId: 'span_id',
  parentSpanId: 'parent_span_id',
  spanKind: 'span_kind',
  spanStatusCode: 'span_status_code',
  spanStatusMessage: 'span_status_message',
  isError: 'is_error',
  userId: 'user_id',
  anonymousId: 'anonymous_id',
  sessionId: 'session_id',
  accountId: 'account_id',
  durationMs: 'duration_ms',
  startTime: 'start_time',
  endTime: 'end_time',
  statusCode: 'status_code',
  loggerName: 'logger.name',
  threadName: 'thread.name',
  metricName: 'metric_name',
  metricType: 'metric_type',
  metricValue: 'metric_value',
  metricUnit: 'metric_unit',
  metricTemporality: 'metric.temporality',
  metricIsMonotonic: 'metric.is_monotonic',
  productId: 'product_id',
  revenueType: 'revenue_type',
  experimentId: 'experiment_id',
  featureFlag: 'feature_flag'
}

export function normalizeCommon(...items: Array<CommonFields | undefined>): JsonObject {
  const output: JsonObject = {}
  for (const item of items) {
    if (!item) continue
    for (const [key, value] of Object.entries(item)) {
      if (value === undefined) continue
      output[fieldMap[key] ?? key] = normalizeJson(value)
    }
  }
  return output
}

export function normalizeJson(value: Json | Date): Json {
  if (value instanceof Date) return value.toISOString()
  if (Array.isArray(value)) return value.map(item => normalizeJson(item))
  if (value && typeof value === 'object') {
    const output: JsonObject = {}
    for (const [key, child] of Object.entries(value)) {
      output[key] = normalizeJson(child as Json)
    }
    return output
  }
  return value
}

export function withoutKeys<T extends Record<string, unknown>>(
  input: T,
  keys: readonly string[]
): CommonFields {
  const skip = new Set(keys)
  const output: CommonFields = {}
  for (const [key, value] of Object.entries(input)) {
    if (!skip.has(key) && isJson(value)) {
      output[key] = value
    }
  }
  return output
}

export function isJson(value: unknown): value is Json {
  if (
    value === null ||
    typeof value === 'boolean' ||
    typeof value === 'number' ||
    typeof value === 'string'
  ) {
    return true
  }
  if (Array.isArray(value)) return value.every(isJson)
  if (value && typeof value === 'object' && value.constructor === Object) {
    return Object.values(value).every(isJson)
  }
  return false
}

File: packages/sdk-py/src/nanotrace/normalize.py
from __future__ import annotations

from datetime import date, datetime, timezone
from typing import Any

from .types import Json, JsonObject

FIELD_ALIASES = (
    ("tenant_id", "tenant_id"),
    ("service.namespace", "service_namespace"),
    ("service.instance.id", "service_instance_id"),
    ("service_version", "service_version"),
    ("host.name", "host_name"),
    ("host.id", "host_id"),
    ("trace_id", "trace_id"),
    ("span_id", "span_id"),
    ("parent_span_id", "parent_span_id"),
    ("span_kind", "span_kind"),
    ("span_status_code", "span_status_code"),
    ("span_status_message", "span_status_message"),
    ("is_error", "is_error"),
    ("user_id", "user_id"),
    ("anonymous_id", "anonymous_id"),
    ("session_id", "session_id"),
    ("account_id", "account_id"),
    ("group_id", "group_id"),
    ("organization_id", "organization_id"),
    ("request_id", "request_id"),
    ("thread_id", "thread_id"),
    ("conversation_id", "conversation_id"),
    ("duration_ms", "duration_ms"),
    ("start_time", "start_time"),
    ("end_time", "end_time"),
    ("status_code", "status_code"),
    ("logger.name", "logger_name"),
    ("thread.name", "thread_name"),
    ("metric_name", "metric_name"),
    ("metric_type", "metric_type"),
    ("metric_value", "metric_value"),
    ("metric_unit", "metric_unit"),
    ("metric.temporality", "metric_temporality"),
    ("metric.is_monotonic", "metric_is_monotonic"),
    ("product_id", "product_id"),
    ("revenue_type", "revenue_type"),
    ("experiment_id", "experiment_id"),
    ("feature_flag", "feature_flag"),
    ("llm.model", "llm_model"),
    ("llm.provider", "llm_provider"),
    ("tool_name", "tool_name"),
)


def _camel_case(value: str) -> str:
    first, *rest = value.split("_")
    return first + "".join(part.capitalize() for part in rest)


FIELD_MAP = {
    alias: canonical
    for canonical, snake_case in FIELD_ALIASES
    for alias in (snake_case, _camel_case(snake_case))
}


def normalize_common(*items: dict[str, Any] | None) -> JsonObject:
    output: JsonObject = {}
    for item in items:
        if not item:
            continue
        for key, value in item.items():
            if value is None:
                continue
            output[FIELD_MAP.get(key, key)] = normalize_json(value)
    return output


def normalize_json(value: Any) -> Json:
    if isinstance(value, datetime):
        return _iso_datetime(value)
    if isinstance(value, date):
        return value.isoformat()
    if isinstance(value, (str, int, float, bool)) or value is None:
        return value
    if isinstance(value, (list, tuple)):
        return [normalize_json(item) for item in value]
    if isinstance(value, dict):
        return {str(key): normalize_json(child) for key, child in value.items()}
    return str(value)


def without_keys(data: dict[str, Any], keys: set[str]) -> dict[str, Any]:
    return {key: value for key, value in data.items() if key not in keys}


def _iso_datetime(value: datetime) -> str:
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.isoformat().replace("+00:00", "Z")
